gensim/models/base_any2vec.py

```diff
@@ -0,0 +1,781 @@
+
+        for thread in workers:
+            thread.daemon = True  # make interrupting the process with ctrl+c easier
+            thread.start()
+
+        trained_word_count, raw_word_count, job_tally = self._log_epoch_progress(
+            progress_queue, job_queue, cur_epoch=cur_epoch, total_examples=total_examples, total_words=total_words,
+            report_delay=report_delay)
+
+        self._set_keyedvectors()
+
+        return trained_word_count, raw_word_count, job_tally
+
+    def train(self, data_iterable, epochs=None, total_examples=None,
+              total_words=None, queue_factor=2, report_delay=1.0, callbacks=(), **kwargs):
+        """Handle multi-worker training."""
+        self._set_train_params(**kwargs)
+        if len(callbacks):
+            self.callbacks = callbacks
+        self.epochs = epochs
+        self._check_training_sanity(
+            epochs=epochs,
+            total_examples=total_examples,
+            total_words=total_words, **kwargs)
+
+        for callback in self.callbacks:
+            callback.on_train_begin(self)
+
+        trained_word_count = 0
+        raw_word_count = 0
+        start = default_timer() - 0.00001
+        job_tally = 0
+
+        for cur_epoch in range(self.epochs):
+            for callback in self.callbacks:
+                callback.on_epoch_begin(self)
+
+            trained_word_count_epoch, raw_word_count_epoch, job_tally_epoch = self._train_epoch(data_iterable,
+                cur_epoch=cur_epoch, total_examples=total_examples, total_words=total_words, queue_factor=queue_factor,
+                report_delay=report_delay)
+            trained_word_count += trained_word_count_epoch
+            raw_word_count += raw_word_count_epoch
+            job_tally += job_tally_epoch
+
+            for callback in self.callbacks:
+                callback.on_epoch_end(self)
+
+        # Log overall time
+        total_elapsed = default_timer() - start
+        self._log_train_end(raw_word_count, trained_word_count, total_elapsed, job_tally)
+
+        self.train_count += 1  # number of times train() has been called
+        self._clear_post_train()
+
+        for callback in self.callbacks:
+            callback.on_train_end(self)
+        return trained_word_count, raw_word_count
+
+    @classmethod
+    def load(cls, fname_or_handle, **kwargs):
+        return super(BaseAny2VecModel, cls).load(fname_or_handle, **kwargs)
+
+    def save(self, fname_or_handle, **kwargs):
+        super(BaseAny2VecModel, self).save(fname_or_handle, **kwargs)
+
+
+class BaseVocabBuilder(utils.SaveLoad):
+    """Class for managing vocabulary of a model. Takes care of building, pruning and updating vocabulary."""
+    def __init__(self):
+        self.vocab = {}
+        self.index2word = []
+
+    def scan_vocab(self, data_iterable, progress_per=10000, **kwargs):
+        """Do an initial scan of all words appearing in data_iterable.
+        Sets num_examples(total examples in data_iterable) and
```

Comment:
Please put spaces in front of opening parentheses `(`, so that it doesn't look like function calls.

Reply:
Do I understand correctly this is meant to set `self.num_examples` and `self.raw_vocab`? If yes, it would be a good idea to create these attributes in the `__init__`, so that they actually exist in the base class. It's OK to initialize to `None`.